midend/hsIndexSimplify.cpp

```diff
@@ -0,0 +1,201 @@
+        } else if (generatedVariables->count(indexString) == 0) {
+            // Generate new temporary variable.
+            auto type = typeMap->getTypeType(arrayIndex->right->type, true);
+            auto name = refMap->newName("hsiVar");
+            auto decl = new IR::Declaration_Variable(name, type);
+            locals->push_back(decl);
+            typeMap->setType(decl, type);
+            newVariable = new IR::PathExpression(arrayIndex->srcInfo, type, new IR::Path(name));
+            generatedVariables->emplace(indexString, newVariable);
+        } else {
+            newVariable = generatedVariables->at(indexString);
+        }
+    }
+}
+
+const IR::Node* HSIndexTransform::postorder(IR::ArrayIndex* curArrayIndex) {
+    if (hsIndexFinder.arrayIndex != nullptr &&
+        curArrayIndex->equiv(*hsIndexFinder.arrayIndex)) {
+        // Translating current array index.
+        auto* newArrayIndex = hsIndexFinder.arrayIndex->clone();
+        newArrayIndex->right = new IR::Constant(newArrayIndex->right->type, index);
+        return newArrayIndex;
+    }
+    return curArrayIndex;
+}
+
+IR::Node* HSIndexSimplifier::eliminateArrayIndexes(HSIndexFinder& aiFinder,
+                                                   IR::Statement* statement,
+                                                   const IR::Expression* expr) {
+    if (aiFinder.arrayIndex == nullptr) {
+        return statement;
+    }
+    const IR::Statement* elseBody = nullptr;
+    auto* curIf = statement->to<IR::IfStatement>();
+    if (curIf != nullptr) {
+        elseBody = curIf->ifFalse;
+    }
+    IR::IndexedVector<IR::StatOrDecl> newComponents;
+    if (aiFinder.newVariable != nullptr) {
+        if (!aiFinder.newVariable->equiv(*aiFinder.arrayIndex->right)) {
+            newComponents.push_back(new IR::AssignmentStatement(aiFinder.arrayIndex->srcInfo,
+            aiFinder.newVariable, aiFinder.arrayIndex->right));
+        }
+    }
+    IR::IfStatement* result = nullptr;
+    IR::IfStatement* curResult = nullptr;
+    IR::IfStatement* newIf = nullptr;
+    const auto* typeStack = aiFinder.arrayIndex->left->type->checkedTo<IR::Type_Stack>();
+    size_t sz = typeStack->getSize();
+    for (size_t i = 0; i < sz; i++) {
+        HSIndexTransform aiRewriter(aiFinder, i);
+        auto* newStatement = statement->apply(aiRewriter)->to<IR::Statement>();
+        auto* newIndex = new IR::Constant(aiFinder.arrayIndex->right->type, i);
+        auto* newCondition = new IR::Equ(aiFinder.newVariable, newIndex);
+        if (curIf != nullptr) {
+            newIf = newStatement->to<IR::IfStatement>()->clone();
+            newIf->condition = new IR::LAnd(newCondition, newIf->condition);
+        } else {
+            newIf = new IR::IfStatement(newCondition, newStatement, nullptr);
+        }
+        if (result == nullptr) {
+            result = newIf;
+            curResult = newIf;
+        } else {
+            curResult->ifFalse = newIf;
```

Comment:
Is there a point to this assignment? curResult is always set to newIf. I think you just need the nullptr check for `.result`

Reply:
I think that it is correct, because we always need to keep innermost if-then-else and result should keep outward if-then-else.

Reply:
`curResult = newIf;`
You have this assignment in both branches, so does `curResult->ifFalse = newIf;` even do anything?

Reply:
curResult->ifFalse = newIf;

is pointless, isn't it?